src/jetstream.rs

```diff
@@ -731,33 +664,692 @@ impl JetStream {
         }
     }
 
+    /// Create an ephemeral push consumer subscription.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let js = nats::jetstream::new(nc);
+    /// let sub = js.subscribe("foo")?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn subscribe(&self, subject: &str) -> io::Result<PushSubscription> {
+        self.do_push_subscribe(subject, None, None)
+    }
+
+    /// Creates a push consumer subscription with options.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use nats::jetstream::{ SubscribeOptions };
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let js = nats::jetstream::new(nc);
+    /// let sub = js.subscribe_with_options("foo", &SubscribeOptions::new()
+    ///     .bind("existing_stream".to_string(), "existing_consumer".to_string())
+    /// )?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn subscribe_with_options(
+        &self,
+        subject: &str,
+        options: &SubscribeOptions,
+    ) -> io::Result<PushSubscription> {
+        self.do_push_subscribe(subject, None, Some(options))
+    }
+
+    /// Creates a push consumer subscription with a queue group.
+    /// The queue group will be used as the durable name.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # fn main() -> std::io::Result<()> {
+    /// # let nc = nats::connect("demo.nats.io")?;
+    /// # let js = nats::jetstream::new(nc);
+    /// let sub = js.queue_subscribe("foo", "production")?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn queue_subscribe(&self, subject: &str, queue: &str) -> io::Result<PushSubscription> {
+        self.do_push_subscribe(subject, Some(queue), None)
+    }
+
+    /// Creates a push consumer subscription with a queue group and options.
+    ///
+    /// If a durable name is not provided, then the queue group will be used as the durable name.
+    ///
+    pub fn queue_subscribe_with_options(
+        &self,
+        subject: &str,
+        queue: &str,
+        options: &SubscribeOptions,
+    ) -> io::Result<PushSubscription> {
+        self.do_push_subscribe(subject, Some(queue), Some(options))
+    }
+
+    fn do_push_subscribe(
+        &self,
+        subject: &str,
+        maybe_queue: Option<&str>,
+        maybe_options: Option<&SubscribeOptions>,
+    ) -> io::Result<PushSubscription> {
+        // If no stream name is specified the subject cannot be empty.
+        if subject.is_empty()
+            && maybe_options
+                .map(|options| options.stream_name.as_ref())
+                .is_none()
+        {
+            return Err(io::Error::new(
```

Comment:
We could probably avoid that runtime error with builder.

Reply:
You're thinking only taking in a single argument which is a builder right?

Yeah I'd want to stick with option variants so the simple "base" signature is the same as it is for NATS subscriptions, e.g `JetStream::subscribe` and `Connection::subscribe` take in the same arguments.